[12.0][IMP] contract: Add contract modifications tracking

victoralmau opened the pull request:
Add contract modification (portal + email + report).

@Tecnativa TT27029

pedrobaeza commented:
Please finish this

victoralmau committed:
[IMP] contract: Add contract modification

pedrobaeza commented on contract/models/contract.py:
```diff
@@ -123,6 +123,41 @@ class ContractContract(models.Model):
+    @api.multi
+    def write(self, vals):
+        res = super(ContractContract, self).write(vals)
+        if 'modification_ids' in vals:
```
You should do this instead on create/write of the subobject. If not, no notification will be sent in case of direct creation through code or specific UI. We have to take into account multi-creation on specific code for sending all the modification in one batch.

pedrobaeza commented on contract/models/contract.py:
```diff
@@ -123,6 +123,41 @@ class ContractContract(models.Model):
+        string='Modifications',
+    )
+
+    @api.multi
```
Don't add this in new code as it's not needed and forbidden in future versions.

pedrobaeza commented on contract/data/mail_template.xml:
```diff
@@ -62,5 +62,26 @@
+        <field name="body_html"><![CDATA[
+<div style="font-family: 'Lucida Grande', Ubuntu, Arial, Verdana, sans-serif; font-size: 12px; color: rgb(34, 34, 34); background-color: #FFF; ">
+    <p>Hello,</p>
+    <p>We have modifications on the contract <strong>${object.name}</strong></p>
```
```suggestion
    <p>We have modifications on the contract <strong>${object.name}</strong> that we want to notify you.</p>
```

pedrobaeza commented on contract/models/contract.py:
```diff
@@ -123,6 +123,41 @@ class ContractContract(models.Model):
+            contract_modification_subtype = self.sudo().env.ref(
+                'contract.mail_message_subtype_contract_modification'
+            )
+            self.message_post_with_template(
```
Don't post messages if no followers.

pedrobaeza commented on contract/models/contract_modification.py:
```diff
@@ -0,0 +1,31 @@
+        ondelete='cascade',
+        index=True
+    )
+    mail_notify = fields.Boolean(
```
Be more precise in the field name. This may mean the contrary: that the notification should be done. Maybe you wanted to say `mail_notified`. Get use to the participle form of the verbs in these cases. Anyway, I think it's better to use `sent` as field name.

pedrobaeza commented on contract/security/ir.model.access.csv:
```diff
@@ -9,3 +9,5 @@
 "contract_line_portal","Recurring portal","model_contract_line","base.group_portal",1,0,0,0
 "contract_template_line_manager","Recurring manager","model_contract_template_line","account.group_account_manager",1,1,1,1
 "contract_template_line_user","Recurring user","model_contract_template_line","account.group_account_invoice",1,0,0,0
+"contract_modification_user","Recurring user","model_contract_modification","account.group_account_invoice",1,1,1,1
```
The name is not very descriptive (although it has been used constantly before). It's better to put:
```suggestion
"contract_modification_user","Contract modifications - User","model_contract_modification","account.group_account_invoice",1,1,1,1
```

pedrobaeza commented on contract/models/contract.py:
```diff
@@ -123,6 +123,41 @@ class ContractContract(models.Model):
+        )
+        if modification_ids_not_notify:
+            template = self.env.ref(
+                'contract.email_contract_modification_template', False
```
You can remove the `False` for not silently fail on this.

pedrobaeza replied:
I'm thinking of using QWeb template instead of this for several reasons:

- Being more hidden from modifications.
- Allow inheritance.

Anyway, the message should be put in the thread. There's an option for using QWeb from mail templates as an envelope, but maybe we can trick it for being the full template and empty inner body.

What do you think?

victoralmau replied:
IMO QWeb template it's a good idea. I will change it now.

pedrobaeza commented on contract/models/contract_modification.py:
```diff
@@ -0,0 +1,31 @@
+        required=True,
+        string='Date'
+    )
+    description = fields.Text(
```
Doing this multi-line, will be ugly when rendering the mail template. Options:

1. Not being multi-line.
2. More elegant: render a table in the mail for displaying this.

victoralmau replied:
Option 2 it's a good idea.

pedrobaeza commented on contract/security/ir.model.access.csv:
```diff
@@ -9,3 +9,5 @@
 "contract_template_line_manager","Recurring manager","model_contract_template_line","account.group_account_manager",1,1,1,1
 "contract_template_line_user","Recurring user","model_contract_template_line","account.group_account_invoice",1,0,0,0
+"contract_modification_user","Recurring user","model_contract_modification","account.group_account_invoice",1,1,1,1
+"contract_modification_portal","Recurring portal","model_contract_modification","base.group_portal",1,0,0,0
```
Same for this:
```suggestion
"contract_modification_portal","Contract modifications - Portal","model_contract_modification","base.group_portal",1,0,0,0
```

pedrobaeza commented on contract/views/contract_portal_templates.xml:
```diff
@@ -165,6 +165,28 @@
                             </table>
                         </div>
+                        <div t-if="contract.modification_ids" id="item_modifications">
+                            <table class="table table-sm" id="sales_order_table">
```
Can this be shown as timeline?

victoralmau replied:
Please, copy some little example about it.

https://github.com/odoo/odoo/blob/14.0/addons/website/views/snippets/s_timeline.xml

pedrobaeza commented on contract/tests/test_contract.py:
```diff
@@ -133,6 +134,36 @@ def _add_template_line(self, overrides=None):
+                )
+            ]
+        })
+        self.assertEqual(len(self.contract.modification_ids.ids), 2)
```
This assert doesn't add value,  as it's testing the framework, not the feature. You can search for mail.message linked to the contract ID if you want to check for the sent mail.

pedrobaeza commented on contract/views/contract.xml:
```diff
@@ -170,6 +170,14 @@
+                                <tree editable="bottom">
+                                    <field name="date"/>
+                                    <field name="description"/>
+                                </tree>
```
Show also the sent flag.

victoralmau commented:
I will add the modifications in the report (PDF).

Tardo commented on contract/models/contract.py:
```diff
@@ -123,6 +123,49 @@ class ContractContract(models.Model):
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        record = super(ContractContract, self).create(vals_list)
```
```suggestion
        record = super().create(vals_list)
```

Tardo commented on contract/models/contract_modification.py:
```diff
@@ -0,0 +1,47 @@
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        record = super(ContractModification, self).create(vals_list)
```
```suggestion
        record = super().create(vals_list)
```

carlosdauden replied:
Multiple records can be returned
```suggestion
        records = super(ContractModification, self).create(vals_list)
```

Tardo commented on contract/models/contract_modification.py:
```diff
@@ -0,0 +1,47 @@
+        return record
+
+    def write(self, vals):
+        res = super(ContractModification, self).write(vals)
```
```suggestion
        res = super().write(vals)
```

Tardo commented on contract/models/contract_modification.py:
```diff
@@ -0,0 +1,47 @@
+    def check_modification_ids_need_sent(self):
+        records_not_sent = self.filtered(lambda x: not x.sent)
+        if records_not_sent:
+            records_not_sent.mapped('contract_id')._modification_mail_send()
```
idk if it's possible, but mapped can return repeated items.

victoralmau replied:
@carlosdauden @pedrobaeza It's possible and need to improve `_modification_mail_send` function to use multi records?

carlosdauden replied:
Char field mapped can return repeated items, but Many2one field mapped retuns a browse with unique items

carlosdauden commented on contract/data/mail_template.xml:
```diff
@@ -62,5 +62,40 @@
+                <tbody>
+                    <t t-foreach="object.modification_ids" t-as="notification_id">
+                        <tr t-if="not notification_id.sent">
+                            <td><span t-esc="notification_id.date" /></td>
+                            <td><t t-esc="notification_id.description" /></td>
```
Why t-esc instead of t-field?

Tardo commented on contract/report/report_contract.xml:
```diff
@@ -99,6 +99,31 @@
+                                <table class="table table-sm">
+                                    <thead>
+                                        <tr>
+                                            <th><strong>Date</strong></th>
```
"TH" should apply bold style ... why need use "strong" here?

carlosdauden commented on contract/report/report_contract.xml:
```diff
@@ -99,6 +99,31 @@
+                                        <t t-foreach="o.modification_ids" t-as="notification_id">
+                                            <tr>
+                                                <td>
+                                                    <span t-esc="notification_id.date"/>
+                                                </td>
+                                                <td class="text-right">
+                                                    <t t-esc="notification_id.description"/>
```
Why t-esc instead of t-field?

Tardo commented on contract/views/assets.xml:
```diff
@@ -7,4 +7,9 @@
     </template>
+    <template id="assets_frontend" inherit_id="web.assets_frontend" name="contract assets">
+        <xpath expr="." position="inside">
+            <link rel="stylesheet" href="/contract/static/src/scss/frontend.scss" type="text/scss"/>
```
The client gets a CSS file
```suggestion
            <link rel="stylesheet" href="/contract/static/src/scss/frontend.scss" type="text/css"/>
```

Tardo reviewed (changes requested):
Some comments are non-blocking ...

pedrobaeza commented on contract/__manifest__.py:
```diff
@@ -17,7 +17,13 @@
+        'account',
+        'product',
+        'portal',
+        'email_template_qweb'
```
This dependency is too hard and can't be added now. When I told you about QWeb, I meant to use core feature of enveloping a QWeb

victoralmau replied:
OK, change now to `message_post_with_view` function

pedrobaeza commented on contract/data/mail_template.xml:
```diff
@@ -62,5 +62,40 @@
+            <table border="1">
+                <thead>
+                    <tr>
+                        <th>State</th>
```
```suggestion
                        <th>Date</th>
```

pedrobaeza reviewed (changes requested):
Please include translations

pedrobaeza commented:
/ocabot merge major

pedrobaeza committed:
[FIX+IMP] contract: Make template to work + other refinements

- Contract modification subtype non default
- Hybrid template between mail.template Jinja + inheritable QWeb
- QWeb element qualified for easing inheritance
- Translations updated
- _modification_mail_send multi-record
- Tests resilience

victoralmau commented:
Sorry Pedro, fix tests and remove `social` in oca dependencies. 
When Travis tests is green, can you launch command again?

victoralmau committed:
[FIX+IMP] contract: Fix tests + Batch modifications

pedrobaeza commented:
/ocabot merge minor
